### PR TITLE
Improve uri.parseQuery to never raise an error

### PR DESCRIPTION
**Pull request by mildred**
In case of malformed query string where there is `=` on the value, handle
this character as part of the value instead of throwing an error.

The following query string should no longer crash a program:

    key=value&key2=x=1

It will be interpreted as:

    [("key", "value"), ("key2", "x=1")]

### CODE REVIEW

**Comment by timotheecour** [comment]
probably correct but please give some links (eg spec, stackoverflow etc) in your PR description to double check;

tests need fixing

File: lib/pure/uri.nim [hunk 1]
```diff
@@ -166,17 +166,20 @@ iterator decodeQuery*(data: string): tuple[key, value: TaintedString] =
   runnableExamples:
     import std/sugar
     let s = collect(newSeq):
```
Comment (timotheecour) [inline lib/pure/uri.nim]: `let s = collect(newSeq):`
=>
`let s = collect:`

is now the preferred way (implies `newSeq`)

**Review by dom96 — CHANGES_REQUESTED** [review]
This is a breaking change. But also whether to raise the error should be customisable.

Please introduce a `strict_parsing` flag like in Python's urllib: https://docs.python.org/3/library/urllib.parse.html#urllib.parse.parse_qs (it should be `true` by default to keep backwards compatibility)

**Comment by timotheecour** [comment]
> This is a breaking change

which code would break?
if you're talking about `doAssertRaises(parseQuery("key=value&key2=x=1"))` I wouldn't categorize it as a breaking change, it'd be something used only in some test suite, not actual code

**Comment by dom96** [comment]
@timotheecour As an example, there may be code out there which relies on being able to check for validity of the query string and explicitly checks for a ``ValueError``.

Are there any downsides to keeping backwards compatibility here?

**Comment by timotheecour** [comment]
if the spec (or most libraries) allows `=`, then I'd much prefer doing a `-d:nimLegacyParseQueryStrict` and being correct by default than legacy behavior by default. The point is to avoid accumulating non-standard behavior over time.

**Comment by mildred** [comment]
I found some references for parsing query strings (and checked them):

- [WhatWG's HTML5 urlencoded parser](https://url.spec.whatwg.org/#concept-urlencoded-parser) accepts literal `=` characters in value
- Go: https://play.golang.org/p/8Yug9Kql2CF no parsing error and `=` interpreted verbatim in value
- Python urllib.parse.parse_qs: same behaviour
- Ruby `CGI::parse` accepts `=` too

**Comment by mildred** [comment]
Updated code with @timotheecour requested changes. Add compilation flag to restore old behaviour suggested. I too prefer the new behaviour by default, it is unlikely to cause breaks in existing apps and the new behaviour is better I believe. it is also conforming to standards and to the general principle of network programming : be liberal in what you accept.

On the contrary, web apps written currently in Nim are probably not handling this specific case of query string parsing, and the new behavior can avoid an error that could in some cases provoke a denial of service.

**Comment by timotheecour** [comment]
Rebase error? Did you rebase against master instead of devel or something?

**Comment by mildred** [comment]
Yep, sorry, updated

File: lib/pure/uri.nim [hunk 2]
```diff
@@ -165,18 +165,21 @@ iterator decodeQuery*(data: string): tuple[key, value: TaintedString] =
...
+      let c = data[result]
+      case c
       of '%': add(field, decodePercent(data, result))
       of '+': add(field, ' ')
-      of '=', '&': break
-      else: add(field, data[result])
+      of '&': break
+      else:
+        if c == sep: break
+        else: add(field, data[result])
```
Comment (timotheecour) [inline lib/pure/uri.nim:183]:
```suggestion
      elif c == sep: break
      else: add(field, data[result])
```

File: lib/pure/uri.nim [hunk 3]
```diff
@@ -165,18 +165,21 @@ iterator decodeQuery*(data: string): tuple[key, value: TaintedString] =
   ## data consists of.
   runnableExamples:
     import std/sugar
-    let s = collect(newSeq):
-      for k, v in decodeQuery("foo=1&bar=2"): (k, v)
-    doAssert s == @[("foo", "1"), ("bar", "2")]
+    let s = collect:
+      for k, v in decodeQuery("foo=1&bar=2=3"): (k, v)
+    doAssert s == @[("foo", "1"), ("bar", "2=3")]
```
Comment (timotheecour) [inline lib/pure/uri.nim]: even simpler:
```nim
import sequtils
doAssert toSeq(decodeQuery("foo=1&bar=2=3")) == @[("foo", "1"), ("bar", "2=3")]
```

Comment (timotheecour) [inline lib/pure/uri.nim]: also add this:
```nim
doAssert toSeq(decodeQuery("&a&=b&=&&")) == @[("", ""), ("a", ""), ("", "b"), ("", ""), ("", "")]
```
which illustrates corner cases

File: lib/pure/uri.nim [hunk 4]
```diff
@@ -165,18 +165,21 @@ iterator decodeQuery*(data: string): tuple[key, value: TaintedString] =
   ## data consists of.
```
Comment (timotheecour) [inline lib/pure/uri.nim]: mention `nimLegacyParseQueryStrict` in doc comment

File: lib/pure/uri.nim [hunk 5]
```diff
@@ -185,16 +188,20 @@ iterator decodeQuery*(data: string): tuple[key, value: TaintedString] =
   # decode everything in one pass:
   while i < data.len:
     setLen(name, 0) # reuse memory
-    i = parseData(data, i, name)
+    i = parseData(data, i, name, '=')
```
Comment (timotheecour) [inline lib/pure/uri.nim:192]: add changelog entry mentioning `nimLegacyParseQueryStrict`

**Review by timotheecour — APPROVED** [review]
LGTM after addressing comments + fixing test failures in tests/stdlib/turi.nim

File: lib/pure/uri.nim (hunk 3 above)
Comment (dom96) [reply lib/pure/uri.nim]: Please don't fill the runnable examples with dozens of corner cases. They're meant to show how to use the procedure, not every single result that is possible.

Put these into a `when isMainModule` or another test file.

Comment (dom96) [reply lib/pure/uri.nim]: We actually have a test file for this already: https://github.com/nim-lang/Nim/blob/devel/tests/stdlib/turi.nim. Please add these tests in there.

**Review by dom96 — CHANGES_REQUESTED** [review]
Hmm. Looking into this a bit more, the specific example we are getting to work here isn't to do with strict parsing, this is in fact fixing a bug in the current parser. Python will parse it fine even with strict parsing enabled:

```
>>> import urllib.parse
>>> urllib.parse.parse_qs('foo=1&bar=2=3')
{'foo': ['1'], 'bar': ['2=3']}
>>> urllib.parse.parse_qs('foo=1&bar=2=3', strict_parsing=True)
{'foo': ['1'], 'bar': ['2=3']}
```

But then why are you changing it to "never raise an error"? Just fix that bug and don't get rid of all errors please. That way this isn't a breaking change.

**Comment by timotheecour** [comment]
> don't get rid of all errors please

which other errors? can you show an input where you expect an error?

File: lib/pure/uri.nim (hunk 3 above)
Comment (timotheecour) [reply lib/pure/uri.nim]: it's not dozens of corner cases, it's literally 2 examples. And yes, when I read docs, I do care about how are corner cases handled, whenever intentional, otherwise you end up doing detective by reading the implementation to figure out the behavior (and need to guess whether future implementations won't break this behavior), which defeats purpose of having runnableExamples in the 1st place.

**Comment by mildred** [comment]
@dom96: I don't understand what you are suggesting... Should we (A) fix the parser to accept the `=` character and never raise an error, not bothering with compatibility define flag or (B) keep the current behaviour and not do anything, suggesting library users to fix their software to avoid the `=` character in the value field.

If we want to be compliant with HTML5 spec, we need to change the behaviour and this is breaking to library users that expects an exception in this special case.

If we want to keep raising an exception, we can't be compliant to the HTML spec.

Note: the `=` character we talk about is the only case when an error is raised. There is no other case where an error is raised as far as I know reading the code.

File: lib/pure/uri.nim (hunk 2 above)
Comment (mildred) [reply lib/pure/uri.nim:183]: This is the else of the case statement, I'm not convinced that elif is a valid option for it.

Comment (timotheecour) [reply lib/pure/uri.nim:183]: it's valid, and it works, i verified

**Comment by mildred** [comment]
Updated the code.

I'm not convinced that the CI errors comes from my code, errors seems to come from fidget and PackageFileParsed (whatever those are) and I don't really see the problem looking at the CI logs...

File: lib/pure/uri.nim [hunk 6]
```diff
@@ -162,21 +162,26 @@ func encodeQuery*(query: openArray[(string, string)], usePlus = true,
...
-  ## data consists of.
+  ## data consists of. If compiled with -d:nimLegacyParseQueryStrict, an error
+  ## is raised when there is an unencoded ``=`` character in a decoded value,
+  ## which was the behavior in Nim < 1.6
   runnableExamples:
     import std/sugar
-    let s = collect(newSeq):
-      for k, v in decodeQuery("foo=1&bar=2"): (k, v)
-    doAssert s == @[("foo", "1"), ("bar", "2")]
+    import sequtils
```
Comment (timotheecour) [inline lib/pure/uri.nim]: `import std/sequtils` works too btw, so you could've written the 1 liner `import std/[sugar,sequtils]`.

However, sugar is no longer used so please remove `import std/sugar`

**Comment by dom96** [comment]
[Looking at the Python implementation](https://github.com/python/cpython/blob/3.9/Lib/urllib/parse.py#L746) it appears this is a case where an exception will be raised with strict parsing on:

```python
>>> urllib.parse.parse_qs('&q', strict_parsing=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.6/urllib/parse.py", line 676, in parse_qs
    max_num_fields=max_num_fields)
  File "/usr/lib/python3.6/urllib/parse.py", line 729, in parse_qsl
    raise ValueError("bad query field: %r" % (name_value,))
ValueError: bad query field: ''
>>> urllib.parse.parse_qs('&q', strict_parsing=False)
{}
```

And you get an exception with decodeQuery for this too.

Another thing to keep in mind here is that cgi makes use of this code, and explicitly checks for exceptions, see the diff that moved this iterator: https://github.com/nim-lang/Nim/commit/689504081f476ec23ebfd7123ad2f8b27c387a97. Are we sure this isn't breaking CGI?

@xflywind any thoughts?

File: lib/pure/uri.nim [hunk 7]
```diff
@@ -162,21 +162,25 @@ func encodeQuery*(query: openArray[(string, string)], usePlus = true,
 
 iterator decodeQuery*(data: string): tuple[key, value: TaintedString] =
   ## Reads and decodes query string ``data`` and yields the (key, value) pairs the
-  ## data consists of.
+  ## data consists of. If compiled with -d:nimLegacyParseQueryStrict, an error
+  ## is raised when there is an unencoded ``=`` character in a decoded value,
+  ## which was the behavior in Nim < 1.6
```
Comment (timotheecour) [inline lib/pure/uri.nim]:
```nim
  ## Reads and decodes query string `data` and yields the `(key, value)` pairs the
  ## data consists of. If compiled with `-d:nimLegacyParseQueryStrict`, an error
  ## is raised when there is an unencoded `=` character in a decoded value,
  ## which was the behavior in Nim < 1.5.1
```

(changed a few things, including https://nim-lang.github.io/Nim/contributing.html
> In nim sources, prefer single backticks

)

**Comment by mildred** [comment]
@dom96 You're right, `cgi` makes use of the same code (and I wanted originally to fix the `cgi.decodeData` iterator. I am removing the error check in the code as this is no longer necessary and updating the changelog.

As for strict checking, I believe this is not the same thing. Is there a use case for struct checking? Nor all languages advertise such a feature. In any case I believe this would belong to another PR and should be clearly defined before any implementation is attempted.

File: lib/pure/uri.nim (hunk 6 above)
Comment (mildred) [reply lib/pure/uri.nim]: done, thanks

File: lib/pure/uri.nim (hunk 7 above)
Comment (mildred) [reply lib/pure/uri.nim]: done

**Comment by dom96** [comment]
Okay, so yes you can make these changes. This code is actually brand new and hasn't been released yet in a Nim version (it was only added 16 days ago), so you don't need to add it to the changelog, nor do you need those `nimLegacyParseQueryStrict` defines.

The only concern I have is with any breakage in cgi, I can see that a test was added to check that the exception is raised correctly in 689504081f476ec23ebfd712.

As for `strict_parsing`, yes we can add it separately. We should really just copy the Python implementation IMO (it includes other things like support for `;` as separators) but that can come in follow up PRs.

**Comment by mildred** [comment]
Well, actually the code in uri right now comes from cgi, it was just moved over. Perhaps the compatibility option should be renamed to refer to cgi instead. I'm not entirely sure it is needed though.

Also, I'm not sure that there is a need for a strict parsing as this does not seems to be a well defined behavior. There is no common implementation of such a strict mode across all languages and I'm not sure copying the Python behavior is of any use on its own. It looks like unneeded complexity for me.

File: lib/pure/uri.nim (hunk 7 above)
Comment (timotheecour) [reply lib/pure/uri.nim]: @mildred thanks; in future, please click "resolve conversation" when something is resolved (unless it's a pushback); thanks!

File: changelog.md [hunk 8]
```diff
@@ -96,6 +96,13 @@
 with other backends. see #9125. Use `-d:nimLegacyJsRound` for previous behavior.
 - Added `socketstream` module that wraps sockets in the stream interface
 
+- Changed the behavior of `uri.decodeQuery` when there are unencoded `=`
```
Comment (timotheecour) [inline changelog.md:99]: > I'm not convinced that the CI errors comes from my code, errors seems to come from fidget and PackageFileParsed (whatever those are) and I don't really see the problem looking at the CI logs...

https://dev.azure.com/nim-lang/255dfe86-e590-40bb-a8a2-3c0295ebdeb1/_apis/build/builds/12009/logs/84

2021-01-11T23:28:40.1530530Z [1m[31mFAIL: [36mtests/stdlib/turi.nim c[0m
...
2021-01-11T23:28:40.1541230Z ../../lib/system/fatal.nim(53, 5) Error: unhandled exception: expected raising 'UriParseError', instead nothing was raised by: 
2021-01-11T23:28:40.1542060Z discard toSeq(decodeQuery("a=1&b=2c=6")) [AssertionDefect]

it comes from this PR, the fix is trivial:
in tests/stdlib/turi.nim, adapt:
```nim
    doAssertRaises(UriParseError):
      discard toSeq(decodeQuery("a=1&b=2c=6"))
```
accordingly

**Comment by timotheecour** [comment]
> Also, I'm not sure that there is a need for a strict parsing as this does not seems to be a well defined behavior.

which of the languages you mentioned in https://github.com/nim-lang/Nim/pull/16647#issuecomment-758054870 define some notion of strict parsing? I'm fine without having such a "strict parsing" option, but if lots of languages have such a notion, then maybe we can reconsider (in future work)

**Comment by mildred** [comment]
> https://github.com/nim-lang/Nim/pull/16647#issuecomment-758298124 which of the languages you mentioned in [#16647 (comment)](https://github.com/nim-lang/Nim/pull/16647#issuecomment-758054870) define some notion of strict parsing?

Of those languages, only [Python](https://docs.python.org/3/library/urllib.parse.html) seems to have strict parsing. [Ruby](https://ruby-doc.org/stdlib-2.5.3/libdoc/cgi/rdoc/CGI.html#method-c-parse) does not neither does [Go](https://golang.org/pkg/net/url/#ParseQuery). Same for the extra options which are only available in Python.

File: changelog.md (hunk 8 above)
Comment (mildred) [reply changelog.md:99]: Thank you, I was looking at the other CI runs where this did not appear, my bad.

Comment (timotheecour) [reply changelog.md:99]: no problem; please click "resolve conversation" for any resolved comment (unless it contains pushback)

**Comment by timotheecour** [comment]
> Of those languages, only Python seems to have strict parsing. Ruby does not neither does Go. Same for the extra options which are only available in Python.

ok, that's convincing. No need to support a notion of strict parsing then.

**Comment by Araq** [comment]
Looks really good this way. Follup PRs can address potential shortcomings that I've missed.

File: lib/pure/cgi.nim [hunk 9]
```diff
@@ -84,23 +84,17 @@ proc getEncodedData(allowedMethods: set[RequestMethod]): string =
 iterator decodeData*(data: string): tuple[key, value: TaintedString] =
   ## Reads and decodes CGI data and yields the (name, value) pairs the
   ## data consists of.
```
Comment (ringabout) [inline lib/pure/cgi.nim:86]: `decodeData` should document `nimLegacyParseQueryStrict` option too.

> And even with `-d:nimLegacyParseQueryStrict`, it never raises `CgiError` now. It actually raises `UriParseError`.

ref https://github.com/nim-lang/Nim/pull/19308#issuecomment-1003737210
